src/@next/components/organisms/CheckoutPayment/CheckoutPayment.tsx

```diff
@@ -0,0 +1,150 @@
+import React, { useEffect, useState } from "react";
+
+import { Checkbox } from "@components/atoms";
+
+import { AddressForm, AddressGridSelector, DiscountForm } from "..";
+import { PaymentGatewaysList } from "../PaymentGatewaysList";
+import * as S from "./styles";
+import { IProps } from "./types";
+
+/**
+ * Payment options used in checkout.
+ */
+const CheckoutPayment: React.FC<IProps> = ({
+  gatewayErrors,
+  billingErrors,
+  selectedUserAddressId,
+  userAddresses,
+  billingAsShippingAddress = false,
+  checkoutBillingAddress,
+  countries,
+  formRef,
+  formId,
+  paymentGateways,
+  setBillingAddress,
+  setBillingAsShippingAddress,
+  promoCodeDiscount,
+  addPromoCode,
+  removeVoucherCode,
+  selectedPaymentGateway,
+  selectedPaymentGatewayToken,
+  selectPaymentGateway,
+  gatewayFormRef,
+  processPayment,
+}: IProps) => {
+  const [showPromoCodeForm, setShowPromoCodeForm] = useState(
+    !!promoCodeDiscount.voucherCode
+  );
+
+  useEffect(() => {
+    const isVoucherCode = !!promoCodeDiscount.voucherCode;
+    if (isVoucherCode) {
+      setShowPromoCodeForm(!!promoCodeDiscount.voucherCode);
+    }
+  }, [promoCodeDiscount.voucherCode]);
+
+  const handleChangeShowPromoCodeForm = () => {
+    if (showPromoCodeForm && promoCodeDiscount?.voucherCode) {
+      removeVoucherCode(promoCodeDiscount?.voucherCode);
+    }
+    setShowPromoCodeForm(!showPromoCodeForm);
+  };
+
+  const adresses =
+    userAddresses
+      ?.filter(function notEmpty<TValue>(
+        value: TValue | null | undefined
+      ): value is TValue {
+        return value !== null && value !== undefined;
+      })
+      .map(address => ({
+        address: {
+          ...address,
+          isDefaultBillingAddress: address.isDefaultBillingAddress || false,
+          isDefaultShippingAddress: address.isDefaultShippingAddress || false,
+          phone: address.phone || undefined,
+        },
+        id: address?.id || "",
+        onSelect: () => null,
+      })) || [];
+
+  return (
+    <S.Wrapper>
+      <S.Section>
+        <S.Title data-cy="checkoutPageSubtitle">BILLING ADDRESS</S.Title>
+        <Checkbox
+          data-cy="checkoutPaymentBillingAsShippingCheckbox"
+          name="billing-same-as-shipping"
+          checked={billingAsShippingAddress}
+          onChange={() =>
+            setBillingAsShippingAddress(!billingAsShippingAddress)
+          }
+        >
+          Same as shipping address
+        </Checkbox>
+        {!billingAsShippingAddress && (
+          <>
+            <S.Divider />
+            {userAddresses ? (
+              <AddressGridSelector
+                addresses={adresses}
+                selectedAddressId={selectedUserAddressId}
+                onSelect={setBillingAddress}
+              />
+            ) : (
+              <AddressForm
+                formId={formId}
+                formRef={formRef}
+                countriesOptions={countries.filter(function notEmpty<TValue>(
```

Comment:
Yes, filtering out null objects in array is required, as graphql schema allows for such null objects - schema should be fixed probably. But in case of moving it to utlis, completely right. It should be there.

Reply:
https://github.com/mirumee/saleor/issues/5515